lib/rust/ensogl/component/grid-view/src/lib.rs

```diff
@@ -147,6 +150,87 @@ impl<E: display::Object> display::Object for VisibleEntry<E> {
 
 
 
+// ====================
+// === ColumnWidths ===
+// ====================
+
+type SegmentTree = segment_tree::SegmentPoint<f32, segment_tree::ops::Add>;
+
+fn segment_tree(width_diffs: Vec<f32>) -> SegmentTree {
+    segment_tree::SegmentPoint::build(width_diffs, segment_tree::ops::Add)
+}
+
+/// Create a new segment tree using an already allocated vector. [`width_diffs`] must have a size
+/// of at least `2N`, where `N` is a count of columns.
+fn segment_tree_noalloc(width_diffs: Vec<f32>) -> SegmentTree {
+    segment_tree::SegmentPoint::build_noalloc(width_diffs, segment_tree::ops::Add)
+}
+
+/// Storage for column widths.
+///
+/// It stores not the actual widths but rather the differences between [`Frp::set_entry_size`] and
+/// the width updated through [`Frp::set_column_width`] or [`EntryFrp::override_column_width`].
+///
+/// It uses a segment tree data structure, allowing efficient access to cumulative differences
+/// across column ranges. We use it in the [`ColumnWidths::pos_offset`] method to calculate the
+/// position of the entry inside the grid view.
+#[derive(Clone, CloneRef, Debug)]
+pub struct ColumnWidths {
+    width_diffs: Rc<RefCell<segment_tree::SegmentPoint<f32, segment_tree::ops::Add>>>,
+}
+
+impl ColumnWidths {
+    /// Constructor. Initializes the storage with every column having the default size inherited
+    /// from [`Frp::set_entry_size`].
+    pub fn new(number_of_columns: usize) -> Self {
+        let init_vec = (0..number_of_columns).map(|_| 0.0).collect();
+        let tree = segment_tree(init_vec);
+        Self { width_diffs: Rc::new(RefCell::new(tree)) }
+    }
+
+    /// Set a width difference for the specified column.
+    pub fn set_width_diff(&self, column: usize, width_diff: f32) {
+        self.width_diffs.borrow_mut().modify(column, width_diff);
+    }
+
+    /// Resize the storage to accommodate the new number of columns.
+    pub fn resize(&self, col_count: usize) {
+        let mut borrowed = self.width_diffs.borrow_mut();
+        let old_length = borrowed.len();
+        let length = col_count;
+        let double_length = length * 2;
+        if old_length < length {
+            let mut new_vec = Vec::with_capacity(double_length);
+            new_vec.resize(double_length, 0.0);
+            new_vec[length..length + old_length].copy_from_slice(borrowed.view());
+            *borrowed = segment_tree_noalloc(new_vec);
+        }
+    }
```

Comment:
Have you seen [`PrefixPoint`](https://docs.rs/segment-tree/2.0.0/segment_tree/struct.PrefixPoint.html) from the same crate? We meet the conditions, and it allows for quick resizing.

Reply:
Replaced with `PrefixPoint`